File: mt4_jog/kinematics.py
"""MT4 forward kinematics and Cartesian jog mixing (matches jog firmware geometry)."""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

# Factory EEPROM geometry (mm). The MT4 is a parallel-link (palletizing) arm:
# J2 sets the upper-arm absolute angle, J3 sets the forearm absolute angle via
# the link rods (independent of J2), and the head platform stays level.
LINKAGE1 = 130.0  # shoulder -> elbow
LINKAGE2 = 150.0  # elbow -> wrist pivot
CENCER_OFFSET = 45.0  # J1 axis -> shoulder, horizontal
CENCER_HEIGHT = 140.0  # shoulder pivot height
HEAD_OFFSET = 35.0  # wrist pivot -> TCP, horizontal (head stays level)
HEAD_HEIGHT = 14.43  # TCP below wrist pivot

# Model angles at the homed pose (firmware step counters = 0): upper arm
# vertical, forearm horizontal — factory firmware reports TCP (230, 0, 255.57).
HOME_J1_DEG = 0.0
HOME_J2_DEG = 90.0
HOME_J3_DEG = 0.0
HOME_J4_DEG = 0.0

# All four measured 2026-07-06 (J2-J4 with a phone clinometer against the
# link; J1 by direct measurement of its yaw rotation), replacing the
# factory-EEPROM-derived guesses -- J1/J2/J3 share a physical motor/gearbox
# design (~35 steps/deg each). J3's own EEPROM setting was missing from the
# dump entirely (the old 35.556 was borrowed from unrelated extra axes), and
# J4's old value (852) was a wrong axis-letter assumption ("d" = J4).
STEPS_PER_DEG: tuple[float, float, float, float] = (35.0, 35.0, 35.0, 45.0)

# +1 if positive step count increases firmware joint angle, -1 if driver is inverted.
# J3 confirmed inverted 2026-07-06: a +299-step probe raised the forearm tip
# instead of lowering it as the old -1.0 sign predicted (photo-confirmed).
J_STEP_SIGN: tuple[float, float, float, float] = (1.0, -1.0, 1.0, 1.0)

DLS_LAMBDA = 0.05

# Positive joint angle => DIR pin low (matches jog_keyboard.py "q" direction).
DIR_POS_HIGH: tuple[bool, bool, bool, bool] = (False, False, False, False)


@dataclass(frozen=True)
class Vec3:
    x: float
    y: float
    z: float

    def normalized(self) -> Vec3:
        n = math.sqrt(self.x * self.x + self.y * self.y + self.z * self.z)
        if n < 1e-9:
            return Vec3(0.0, 0.0, 0.0)
        return Vec3(self.x / n, self.y / n, self.z / n)


@dataclass(frozen=True)
class JointAnglesDeg:
    j1: float
    j2: float
    j3: float
    j4: float = 0.0

    @classmethod
    def from_steps(cls, steps: tuple[int, int, int, int]) -> JointAnglesDeg:
        spd = STEPS_PER_DEG
        s = J_STEP_SIGN
        return cls(
            HOME_J1_DEG + s[0] * steps[0] / spd[0],
            HOME_J2_DEG + s[1] * steps[1] / spd[1],
            HOME_J3_DEG + s[2] * steps[2] / spd[2],
            HOME_J4_DEG + s[3] * steps[3] / spd[3],
        )


def fk_tcp(q: JointAnglesDeg) -> Vec3:
    q1, q2, q3 = map(math.radians, (q.j1, q.j2, q.j3))
    radial = CENCER_OFFSET + LINKAGE1 * math.cos(q2) + LINKAGE2 * math.cos(q3) + HEAD_OFFSET
    return Vec3(
        radial * math.cos(q1),
        radial * math.sin(q1),
        CENCER_HEIGHT + LINKAGE1 * math.sin(q2) + LINKAGE2 * math.sin(q3) - HEAD_HEIGHT,
    )


def jacobian_mm_per_deg(q: JointAnglesDeg) -> np.ndarray:
    """3x4 matrix: d(TCP mm)/d(joint deg)."""
    j = np.zeros((3, 4))
    base = np.array([q.j1, q.j2, q.j3, q.j4], dtype=float)
    tcp0 = fk_tcp(q)
    p0 = np.array([tcp0.x, tcp0.y, tcp0.z])
    for i in range(4):
        trial = base.copy()
        trial[i] += 0.1
        q_t = JointAnglesDeg(*trial)
        p1 = np.array([fk_tcp(q_t).x, fk_tcp(q_t).y, fk_tcp(q_t).z])
        j[:, i] = (p1 - p0) / 0.1
    return j


def cartesian_joint_rates_deg(
    q: JointAnglesDeg,
    direction: Vec3,
    *,
    hold_orientation: bool = True,
) -> tuple[float, float, float, float] | None:
    """Model-space joint deg rates for unit world velocity along direction.

    When hold_orientation, J4 counters base yaw 1:1 (dq4 = -dq1), matching the
    firmware's `orient on` behavior. Driver step signs are applied in
    cartesian_step_rates().
    """
    v = direction.normalized()
    if abs(v.x) + abs(v.y) + abs(v.z) < 1e-9:
        return None

    j_pos = jacobian_mm_per_deg(q)[:, :3]
    jjt = j_pos @ j_pos.T + (DLS_LAMBDA**2) * np.eye(3)
    try:
        y = np.linalg.solve(jjt, np.array([v.x, v.y, v.z]))
        dq123 = j_pos.T @ y
    except np.linalg.LinAlgError:
        return None

    dq4 = 0.0
    if hold_orientation and abs(dq123[0]) > 1e-6:
        dq4 = -dq123[0]

    return float(dq123[0]), float(dq123[1]), float(dq123[2]), dq4


def cartesian_step_rates(
    q: JointAnglesDeg,
    direction: Vec3,
    *,
    hold_orientation: bool = True,
    master_scale: int = 10_000,
) -> tuple[int, int, int, int, int] | None:
    """Bresenham integer rates (j1..j4, master) for firmware cj command."""
    rates = cartesian_joint_rates_deg(q, direction, hold_orientation=hold_orientation)
    if rates is None:
        return None

    steps = [J_STEP_SIGN[i] * rates[i] * STEPS_PER_DEG[i] for i in range(4)]

    # Peak/master-scale spans all four joints. This used to be a problem
    # when J4's steps/deg (852, an axis-letter misassignment) was ~19x J1's,
    # letting a modest orientation-hold correction dominate the DDA timing
    # budget. Now that J4 is correctly calibrated (~45, close to J1's ~44),
    # including it here costs at most a few percent of speed and gives exact
    # wrist-unwind fidelity instead of clamping J4 short.
    peak = max(abs(s) for s in steps)
    if peak < 1e-9:
        return None

    scale = master_scale / peak
    ints = [int(round(s * scale)) for s in steps]
    return ints[0], ints[1], ints[2], ints[3], master_scale


def steps_to_dir_high(joint_index: int, rate: int) -> bool:
    positive = rate > 0
    return DIR_POS_HIGH[joint_index] if positive else not DIR_POS_HIGH[joint_index]


# ---------------------------------------------------------------------------
# Position-level IK (closed form) -- for bounded point-to-point moves via the
# firmware `m` command, as opposed to the resolved-rate jog IK above.
# ---------------------------------------------------------------------------


def _wrap_deg(a: float) -> float:
    return (a + 180.0) % 360.0 - 180.0


def ik_q2_q3(
    radial: float, z: float, near_q2: float, near_q3: float
) -> tuple[float, float] | None:
    """Closed-form two-link solve in the arm's vertical plane:

        LINKAGE1*(cos q2, sin q2) + LINKAGE2*(cos q3, sin q3) = target

    (absolute joint angles; circle-circle intersection). Returns the branch
    nearest (near_q2, near_q3), or None if the target is out of reach.
    """
    tx = radial - CENCER_OFFSET - HEAD_OFFSET
    ty = z - CENCER_HEIGHT + HEAD_HEIGHT
    d = math.hypot(tx, ty)
    if (
        d < 1e-6
        or d > LINKAGE1 + LINKAGE2
        or d < abs(LINKAGE1 - LINKAGE2)
    ):
        return None

    cos_alpha = (LINKAGE1**2 + d * d - LINKAGE2**2) / (2 * LINKAGE1 * d)
    cos_alpha = max(-1.0, min(1.0, cos_alpha))
    alpha = math.acos(cos_alpha)
    beta = math.atan2(ty, tx)

    best: tuple[float, float, float] | None = None
    for sign in (1.0, -1.0):
        q2 = math.degrees(beta + sign * alpha)
        p1x = LINKAGE1 * math.cos(math.radians(q2))
        p1y = LINKAGE1 * math.sin(math.radians(q2))
        q3 = math.degrees(math.atan2(ty - p1y, tx - p1x))
        dist = abs(_wrap_deg(q2 - near_q2)) + abs(_wrap_deg(q3 - near_q3))
        if best is None or dist < best[0]:
            best = (dist, q2, q3)
    assert best is not None
    return best[1], best[2]


def ik_position(
    x: float,
    y: float,
    z: float,
    *,
    near: JointAnglesDeg,
    hold_orientation: bool = True,
) -> JointAnglesDeg | None:
    """Full position IK: TCP (x, y, z) mm -> joint angles, elbow branch and
    J1 wrap chosen nearest `near`. When hold_orientation, J4 counters the J1
    change 1:1 (same convention as the jog IK); pass False to leave J4
    untouched.
    """
    q1 = math.degrees(math.atan2(y, x))
    q1 = near.j1 + _wrap_deg(q1 - near.j1)
    sol = ik_q2_q3(math.hypot(x, y), z, near.j2, near.j3)
    if sol is None:
        return None
    q4 = near.j4 - (q1 - near.j1) if hold_orientation else near.j4
    return JointAnglesDeg(q1, sol[0], sol[1], q4)


def steps_from_angles(q: JointAnglesDeg) -> tuple[int, int, int, int]:
    """Inverse of JointAnglesDeg.from_steps(): absolute step counters."""
    home = (HOME_J1_DEG, HOME_J2_DEG, HOME_J3_DEG, HOME_J4_DEG)
    vals = (q.j1, q.j2, q.j3, q.j4)
    return tuple(
        int(round((vals[i] - home[i]) * STEPS_PER_DEG[i] * J_STEP_SIGN[i]))
        for i in range(4)
    )  # type: ignore[return-value]

File: firmware/mt4_jog/src/kinematics.cpp
#include "kinematics.h"

#include <math.h>
#include <stdio.h>

/* The MT4 is a parallel-link (palletizing) arm:
 *  - J2 sets the upper-arm absolute angle q2 (from horizontal).
 *  - J3 sets the forearm absolute angle q3 via the link rods, so the forearm
 *    does NOT rotate when J2 moves.
 *  - The head platform stays level: HEAD_OFFSET is horizontal (radially out),
 *    HEAD_HEIGHT drops the TCP below the wrist pivot.
 * Home pose q = (0, 103, 4.7): measured directly (see kinematics.h), not the
 * upper-arm-vertical/forearm-horizontal (0, 90, 0) previously assumed.
 */
static const float LINKAGE1 = 130.0f;      /* shoulder -> elbow */
static const float LINKAGE2 = 150.0f;      /* elbow -> wrist pivot */
static const float CENCER_OFFSET = 45.0f;  /* J1 axis -> shoulder, horizontal */
static const float CENCER_HEIGHT = 140.0f; /* shoulder pivot height */
static const float HEAD_OFFSET = 35.0f;    /* wrist pivot -> TCP, horizontal */
static const float HEAD_HEIGHT = 14.43f;   /* TCP below wrist pivot */

/* All four measured 2026-07-06 (J2-J4 with a phone clinometer against the
 * link; J1 by direct measurement of its yaw rotation), replacing the
 * factory-EEPROM-derived guesses -- J1/J2/J3 share a physical motor/gearbox
 * design (~35 steps/deg each). J3's own EEPROM setting was missing from the
 * dump entirely (the old 35.556 was borrowed from unrelated extra axes),
 * and J4's old value (852) was a wrong axis-letter assumption ("d" = J4). */
const float MT4_STEPS_PER_DEG[MT4_NUM_JOINTS] = {35.0f, 35.0f, 35.0f,
                                                 45.0f};
static const float J_STEP_SIGN[MT4_NUM_JOINTS] = {
    MT4_J1_STEP_SIGN, MT4_J2_STEP_SIGN, MT4_J3_STEP_SIGN, MT4_J4_STEP_SIGN};
static const float DLS_LAMBDA = 0.05f;

void mt4_fk_tcp(const JointAnglesDeg *q, Vec3 *out) {
  const float q1 = q->j1 * (float)(M_PI / 180.0);
  const float q2 = q->j2 * (float)(M_PI / 180.0);
  const float q3 = q->j3 * (float)(M_PI / 180.0);

  const float radial = CENCER_OFFSET + LINKAGE1 * cosf(q2) +
                       LINKAGE2 * cosf(q3) + HEAD_OFFSET;
  out->x = radial * cosf(q1);
  out->y = radial * sinf(q1);
  out->z = CENCER_HEIGHT + LINKAGE1 * sinf(q2) + LINKAGE2 * sinf(q3) -
           HEAD_HEIGHT;
}

static void jacobian_mm_per_deg(const JointAnglesDeg *q, float j[3][MT4_NUM_JOINTS]) {
  Vec3 p0;
  JointAnglesDeg trial = *q;
  mt4_fk_tcp(&trial, &p0);

  for (uint8_t i = 0; i < MT4_NUM_JOINTS; ++i) {
    float *slot = &((float *)&trial.j1)[i];
    const float saved = *slot;
    *slot = saved + 0.1f;
    Vec3 p1;
    mt4_fk_tcp(&trial, &p1);
    *slot = saved;

    j[0][i] = (p1.x - p0.x) / 0.1f;
    j[1][i] = (p1.y - p0.y) / 0.1f;
    j[2][i] = (p1.z - p0.z) / 0.1f;
  }
}

static bool solve_dls3(const float j[3][3], const float v[3], float dq[3]) {
  float jjt[3][3];
  for (uint8_t r = 0; r < 3; ++r) {
    for (uint8_t c = 0; c < 3; ++c) {
      float sum = 0.0f;
      for (uint8_t k = 0; k < 3; ++k) {
        sum += j[r][k] * j[c][k];
      }
      jjt[r][c] = sum;
    }
  }

  const float lambda2 = DLS_LAMBDA * DLS_LAMBDA;
  for (uint8_t i = 0; i < 3; ++i) {
    jjt[i][i] += lambda2;
  }

  float m[3][4];
  for (uint8_t i = 0; i < 3; ++i) {
    for (uint8_t k = 0; k < 3; ++k) {
      m[i][k] = jjt[i][k];
    }
    m[i][3] = v[i];
  }

  for (uint8_t col = 0; col < 3; ++col) {
    uint8_t pivot = col;
    float best = fabsf(m[col][col]);
    for (uint8_t r = col + 1; r < 3; ++r) {
      const float val = fabsf(m[r][col]);
      if (val > best) {
        best = val;
        pivot = r;
      }
    }
    if (best < 1e-6f) {
      return false;
    }
    if (pivot != col) {
      for (uint8_t k = col; k < 4; ++k) {
        const float tmp = m[col][k];
        m[col][k] = m[pivot][k];
        m[pivot][k] = tmp;
      }
    }

    const float div = m[col][col];
    for (uint8_t k = col; k < 4; ++k) {
      m[col][k] /= div;
    }
    for (uint8_t r = 0; r < 3; ++r) {
      if (r == col) {
        continue;
      }
      const float factor = m[r][col];
      for (uint8_t k = col; k < 4; ++k) {
        m[r][k] -= factor * m[col][k];
      }
    }
  }

  const float y[3] = {m[0][3], m[1][3], m[2][3]};
  for (uint8_t c = 0; c < 3; ++c) {
    float sum = 0.0f;
    for (uint8_t r = 0; r < 3; ++r) {
      sum += j[r][c] * y[r];
    }
    dq[c] = sum;
  }
  return true;
}

bool mt4_cartesian_rates(const JointAnglesDeg *q, const Vec3 *dir_unit,
                         bool hold_orient, CartesianRates *out) {
  float j[3][MT4_NUM_JOINTS];
  jacobian_mm_per_deg(q, j);

  float a[3][3];
  for (uint8_t r = 0; r < 3; ++r) {
    for (uint8_t c = 0; c < 3; ++c) {
      a[r][c] = j[r][c];
    }
  }

  const float b[3] = {dir_unit->x, dir_unit->y, dir_unit->z};
  float dq[3];
  if (!solve_dls3(a, b, dq)) {
    return false;
  }

  /* dq is in model-angle space; wrist unwind counters base yaw 1:1 before the
   * per-driver step signs are applied (real J1/J4 axis alignment and any
   * mechanical wrist coupling are assumed exact). */
  float dq4 = 0.0f;
  if (hold_orient && fabsf(dq[0]) > 1e-4f) {
    dq4 = -dq[0];
  }

  const float steps[4] = {
      J_STEP_SIGN[0] * dq[0] * MT4_STEPS_PER_DEG[0],
      J_STEP_SIGN[1] * dq[1] * MT4_STEPS_PER_DEG[1],
      J_STEP_SIGN[2] * dq[2] * MT4_STEPS_PER_DEG[2],
      J_STEP_SIGN[3] * dq4 * MT4_STEPS_PER_DEG[3]};

  /* Peak/master-scale spans all four joints. This used to be a problem when
   * J4's steps/deg (852, an axis-letter misassignment -- see kinematics.h)
   * was ~19x J1's, letting a modest orientation-hold correction dominate the
   * DDA timing budget and throttle the primary motion to a crawl. Now that
   * J4 is correctly calibrated (~45, close to J1's ~44), including it here
   * costs at most a few percent of speed and gives exact wrist-unwind
   * fidelity instead of clamping J4 short. */
  float peak = 0.0f;
  for (uint8_t i = 0; i < MT4_NUM_JOINTS; ++i) {
    const float v = fabsf(steps[i]);
    if (v > peak) {
      peak = v;
    }
  }
  if (peak < 1e-6f) {
    return false;
  }

  const float scale = (float)MT4_CJ_MASTER / peak;
  out->j1 = (int32_t)lroundf(steps[0] * scale);
  out->j2 = (int32_t)lroundf(steps[1] * scale);
  out->j3 = (int32_t)lroundf(steps[2] * scale);
  out->j4 = (int32_t)lroundf(steps[3] * scale);
  out->master = MT4_CJ_MASTER;
  return true;
}

static float wrap_deg(float a) {
  a = fmodf(a + 180.0f, 360.0f);
  if (a < 0.0f) {
    a += 360.0f;
  }
  return a - 180.0f;
}

/* Closed-form two-link solve in the arm's vertical plane (circle-circle
 * intersection): LINKAGE1*(cos q2, sin q2) + LINKAGE2*(cos q3, sin q3) =
 * (tx, ty). Picks the branch nearest (near_q2, near_q3). Mirrors
 * mt4_jog/kinematics.py's ik_q2_q3(). */
static bool ik_q2_q3(float radial, float z, float near_q2, float near_q3,
                     float *out_q2, float *out_q3) {
  const float tx = radial - CENCER_OFFSET - HEAD_OFFSET;
  const float ty = z - CENCER_HEIGHT + HEAD_HEIGHT;
  const float d = sqrtf(tx * tx + ty * ty);
  if (d < 1e-6f || d > LINKAGE1 + LINKAGE2 || d < fabsf(LINKAGE1 - LINKAGE2)) {
    return false;
  }

  float cos_alpha = (LINKAGE1 * LINKAGE1 + d * d - LINKAGE2 * LINKAGE2) /
                    (2.0f * LINKAGE1 * d);
  cos_alpha = fmaxf(-1.0f, fminf(1.0f, cos_alpha));
  const float alpha = acosf(cos_alpha);
  const float beta = atan2f(ty, tx);
  const float rad2deg = (float)(180.0 / M_PI);
  const float deg2rad = (float)(M_PI / 180.0);

  bool have_best = false;
  float best_dist = 0.0f;
  float best_q2 = 0.0f;
  float best_q3 = 0.0f;
  for (int8_t sign = -1; sign <= 1; sign += 2) {
    const float q2 = (beta + (float)sign * alpha) * rad2deg;
    const float p1x = LINKAGE1 * cosf(q2 * deg2rad);
    const float p1y = LINKAGE1 * sinf(q2 * deg2rad);
    const float q3 = atan2f(ty - p1y, tx - p1x) * rad2deg;
    const float dist =
        fabsf(wrap_deg(q2 - near_q2)) + fabsf(wrap_deg(q3 - near_q3));
    if (!have_best || dist < best_dist) {
      have_best = true;
      best_dist = dist;
      best_q2 = q2;
      best_q3 = q3;
    }
  }
  *out_q2 = best_q2;
  *out_q3 = best_q3;
  return true;
}

bool mt4_ik_position(const JointAnglesDeg *near, float x, float y, float z,
                     float j4_deg, JointAnglesDeg *out) {
  const float rad2deg = (float)(180.0 / M_PI);
  float q1 = atan2f(y, x) * rad2deg;
  q1 = near->j1 + wrap_deg(q1 - near->j1);

  const float radial = sqrtf(x * x + y * y);
  float q2 = 0.0f;
  float q3 = 0.0f;
  if (!ik_q2_q3(radial, z, near->j2, near->j3, &q2, &q3)) {
    return false;
  }

  out->j1 = q1;
  out->j2 = q2;
  out->j3 = q3;
  out->j4 = j4_deg;
  return true;
}
